README.md
Krypt: Simple, Secure, Symmetric Encryption
=====

Krypt uses encryption best practices to produce very secure, portable JSON objects using a simple shared key.

## Install

```
npm install -g krypt
```

## CLI Usage

### Encryption

```
krypt --encrypt ./path/to/file.json --secret someSecretValue --out ./path/to/new/encrypted.out
```

### Decryption

```
krypt --decrypt ./path/to/new/encrypted.out --secret someSecretValue --out ./path/to/file.json
```

## Module Usage

### Encryption

```
var krypt = require('krypt');

var encryptedSync = krypt.encrypt('Input Value', 'someSecretValue');

krypt.encrypt(encrypted, 'someSecretValue', function (err, encryptedAsync) {
  // async encrypt
});
```

### Decryption

```
var krypt = require('krypt');

var plainTextSync = krypt.decrypt(encrypted, 'someSecretValue');

krypt.decrypt(encrypted, 'someSecretValue', function (err, plainTextAsync) {
  // async decrypt
});
```

## Algorithms

Krypt tries to incorporate current encryption best practices into a simple to use module. As such, Krypt uses the following:

+ AES 256
+ CBC
+ Key Stretching w/ PBKDF2 @ 128,000 iterations (default)
+ Random IV / encrypted value
+ Random salt / encrypted value
+ sha512 digest


CHANGELOG.md
## 0.1.0 (3/7/2014)

+ Initial release

## 0.2.0 (3/7/2014)

+ Updated defaults for key length / stretching iterations
+ Enhanced the output to be portable across future versions of Krypt

## 0.2.1 (3/7/2014)

+ Fixed module export

## 0.2.2 (3/7/2014)

+ Fixed typo in payload

## 0.3.0 (3/18/2014)

+ Updated `keyLength` to be expressed in bits (w/ legacy support for converting 32 Byte length keys)
+ Added support to inject custom context into the encrypted result
  + Useful for recording metadata, such as the name of the key you are using, etc.

## 0.4.0 (7/17/2015)

+ Support async encryption / decryption thanks to Spencer Rinehart (@nubs)
  + Introduction of explicit `encryptSync`, `encryptAsync` and `decryptSync` and `decryptAsync` methods.
  + `encrypt` and `decrypt` will automatically choose async or sync based on the presence of a callback.

## 0.5.0 (2/25/2018)

+ Support for Node 6+
+ Dropping official support for Node 0.12
+ Added support for configuring the digest
+ Updated defaults
  + sha512
  + 128,000 iterations


test/krypt.js
var chai = require('chai'),
    expect = chai.expect,
    krypt = require('../lib/krypt');

var PLAIN_TEXT = 'This is my text!',
    SECRET = 'A59C60BCF422D';

describe('Krypt', function () {

  describe('synchronous functions', function () {

    it('should successfuly encrypt a value', function (done) {

      var encrypted = krypt.encryptSync(PLAIN_TEXT, SECRET);
      expect(encrypted).to.have.property('iv');
      expect(encrypted).to.have.property('salt');
      expect(encrypted).to.have.property('digest');
      expect(encrypted).to.have.property('value');
      done();

    });

    it('should successfuly decrypt a JSON value', function (done) {

      var encrypted = krypt.encryptSync(PLAIN_TEXT, SECRET),
          decrypted = krypt.decryptSync(encrypted, SECRET);

      expect(encrypted.digest).to.equal(krypt.DEFAULT_DIGEST);
      expect(decrypted).to.deep.equal(PLAIN_TEXT);
      done();

    });

    it('should successfuly decrypt a string value', function (done) {

      var encrypted = krypt.encryptSync(PLAIN_TEXT, SECRET),
          decrypted = krypt.decryptSync(JSON.stringify(encrypted), SECRET);

      expect(decrypted).to.deep.equal(PLAIN_TEXT);
      done();

    });

    it('should successfuly decrypt with the deprecated default digest', function (done) {
      // Change to deprecated default
      krypt.setDigest(krypt.DEFAULT_DEPRECATED_DIGEST);

      var encrypted = krypt.encryptSync(PLAIN_TEXT, SECRET);

      // Test and Remove the digest to ensure backwards compatibility
      expect(encrypted.digest).to.equal(krypt.DEFAULT_DEPRECATED_DIGEST);
      delete encrypted.digest;
      expect(encrypted.digest).to.not.exist;

      // Reset to the updated default digest
      krypt.setDigest(krypt.DEFAULT_DIGEST);

      var decrypted = krypt.decryptSync(JSON.stringify(encrypted), SECRET);

      expect(decrypted).to.deep.equal(PLAIN_TEXT);
      done();

    });

    it('should fail to encrypt a string value without input', function (done) {

      var encrypted, err;

      try {
        encrypted = krypt.encryptSync(null, SECRET);
      } catch (e) {
        err = e;
      }

      expect(encrypted).to.not.exist;
      expect(err).to.exist;
      expect(err.message).to.deep.equal('You must provide a value to encrypt');
      done();
    });

    it('should fail to encrypt a string value without a secret', function (done) {

      var encrypted, err;

      try {
        encrypted = krypt.encryptSync(PLAIN_TEXT);
      } catch (e) {
        err = e;
      }

      expect(encrypted).to.not.exist;
      expect(err).to.exist;
      expect(err.message).to.deep.equal('A \'secret\' is required to encrypt');
      done();
    });


    it('should fail to decrypt a string value without input', function (done) {

      var encrypted = krypt.encryptSync(PLAIN_TEXT, SECRET),
          decrypted, err;

      try {
        decrypted = krypt.decryptSync(null, SECRET);
      } catch (e) {
        err = e;
      }

      expect(decrypted).to.not.exist;
      expect(err).to.exist;
      expect(err.message).to.deep.equal('You must provide a value to decrypt');
      done();
    });

    it('should fail to decrypt a string value without a secret', function (done) {

      var encrypted = krypt.encryptSync(PLAIN_TEXT, SECRET),
          decrypted, err;

      try {
        decrypted = krypt.decryptSync(JSON.stringify(encrypted));
      } catch (e) {
        err = e;
      }

      expect(decrypted).to.not.exist;
      expect(err).to.exist;
      expect(err.message).to.deep.equal('A \'secret\' is required to decrypt');
      done();
    });

  });

  describe('asynchronous functions', function () {

    it('should successfuly encrypt a value', function (done) {

      krypt.encryptAsync(PLAIN_TEXT, SECRET, function(err, encrypted) {
        expect(err).to.be.null;
        expect(encrypted).to.have.property('iv');
        expect(encrypted).to.have.property('salt');
        expect(encrypted).to.have.property('value');
        done();
      });

    });

    it('should successfuly decrypt a JSON value', function (done) {

      var encrypted = krypt.encryptSync(PLAIN_TEXT, SECRET);

      krypt.decryptAsync(encrypted, SECRET, function(err, decrypted) {
        expect(err).to.be.null;
        expect(decrypted).to.deep.equal(PLAIN_TEXT);
        done();
      });

    });

    it('should successfuly decrypt a string value', function (done) {

      var encrypted = krypt.encryptSync(PLAIN_TEXT, SECRET);

      krypt.decryptAsync(JSON.stringify(encrypted), SECRET, function(err, decrypted) {
        expect(err).to.be.null;
        expect(decrypted).to.deep.equal(PLAIN_TEXT);
        done();
      });

    });

    it('should successfuly decrypt a JSON value with the deprecated default digest', function (done) {
      krypt.setDigest(krypt.DEFAULT_DEPRECATED_DIGEST);

      var encrypted = krypt.encryptSync(PLAIN_TEXT, SECRET);

      // Test and Remove the digest to ensure backwards compatibility
      expect(encrypted.digest).to.equal(krypt.DEFAULT_DEPRECATED_DIGEST);
      delete encrypted.digest;
      expect(encrypted.digest).to.not.exist;

      krypt.decryptAsync(encrypted, SECRET, function(err, decrypted) {
        expect(err).to.be.null;
        expect(decrypted).to.deep.equal(PLAIN_TEXT);
        done();
      });

    });

    it('should fail to encrypt a string value without input', function (done) {

      krypt.encryptAsync(null, SECRET, function (err, encrypted) {
        expect(encrypted).to.not.exist;
        expect(err).to.exist;
        expect(err.message).to.deep.equal('You must provide a value to encrypt');
        done();
      });

    });

    it('should fail to encrypt a string value without a secret', function (done) {

      krypt.encryptAsync(PLAIN_TEXT, null, function (err, encrypted) {
        expect(encrypted).to.not.exist;
        expect(err).to.exist;
        expect(err.message).to.deep.equal('A \'secret\' is required to encrypt');
        done();
      });

    });


    it('should fail to decrypt a string value without input', function (done) {

      var encrypted = krypt.encryptSync(PLAIN_TEXT, SECRET);

      krypt.decryptAsync(null, SECRET, function (err, decrypted) {
        expect(decrypted).to.not.exist;
        expect(err).to.exist;
        expect(err.message).to.deep.equal('You must provide a value to decrypt');
        done();
      });
    });

    it('should fail to decrypt a string value without a secret', function (done) {

      var encrypted = krypt.encryptSync(PLAIN_TEXT, SECRET);

      krypt.decryptAsync(JSON.stringify(encrypted), null, function (err, decrypted) {
        expect(decrypted).to.not.exist;
        expect(err).to.exist;
        expect(err.message).to.deep.equal('A \'secret\' is required to decrypt');
        done();
      });
    });

  });

});


index.js
#!/usr/bin/env node

/*!
 * Krypt
 * Copyright(c) 2014 Mike Moulton
 */

'use strict';

var fs = require('fs'),
    path = require('path'),
    krypt = require('./lib/krypt');


module.exports = krypt;


/**
 * CLI
 */
function main() {

  var nconf = require('nconf');

  // Load command line options
  nconf
    .argv({
      'encrypt': {
        short: 'e',
        describe: 'Path of file to encrypt'
      },
      'decrypt': {
        short: 'd',
        describe: 'Path of file to decrypt'
      },
      'secret': {
        short: 's',
        describe: 'Secret used to encrypt/decrypt file',
        demand: true
      },
      'iterations': {
        short: 'i',
        describe: 'Iterations to use for key stretching',
        default: 64000
      },
      'digest': {
        short: 'd',
        describe: 'HMAC digest algorithm to use for deriving the key',
        default: 'sha512'
      },
      'length': {
        short: 'l',
        describe: 'Length of the key to use in Bytes',
        default: 32
      },
      'out': {
        short: 'o',
        describe: 'File location to write result to'
      }
    });

  var secret = nconf.get('secret');

  if (nconf.get('iterations')) {
    krypt.setIterations(nconf.get('iterations'));
  }

  if (nconf.get('length')) {
    krypt.setKeyLength(nconf.get('length'));
  }

  if (nconf.get('encrypt')) {

    try {
      var encryptInput = fs.readFileSync(path.resolve(nconf.get('encrypt')), 'utf8'),
          encrypted = krypt.encrypt(encryptInput, secret);

      if (nconf.get('out')) {
        fs.writeFileSync(
          path.join(__dirname, nconf.get('out')),
          JSON.stringify(encrypted, null, '  '),
          'utf8');
      } else {
        console.log(JSON.stringify(encrypted, null, '  '));
      }
    } catch (err) {
      console.error(err);
      process.exit(1);
    }

  } else if (nconf.get('decrypt')) {

    try {
      var decryptInput = fs.readFileSync(path.resolve(nconf.get('decrypt')), 'utf8'),
          decrypted = krypt.decrypt(decryptInput, secret);

      if (nconf.get('out')) {
        fs.writeFileSync(
          path.join(__dirname, nconf.get('out')),
          decrypted,
          'utf8');
      } else {
        console.log(decrypted);
      }
    } catch (err) {
      console.error(err);
      process.exit(1);
    }

  }

}

// If run from the command line, start application
if (require.main === module) {
  main();
}


lib/krypt.js
/*!
 * Krypt
 * Copyright(c) 2014 Mike Moulton
 */

'use strict';

var crypto = require('crypto'),
    _ = require('lodash');

var CIPHER = 'aes-256-cbc',
    KEY_DERIVATION = 'pbkdf2',
    DEFAULT_KEY_LENGTH = 256,
    DEFAULT_ITERATIONS = 128000,
    DEFAULT_DIGEST = 'sha512',
    DEFAULT_DEPRECATED_DIGEST = 'sha1';


module.exports = new Krypt();
exports.Krypt = Krypt;


function Krypt(config) {
  if (!(this instanceof Krypt)) {
    return new Krypt(config);
  }

  config = config || {};

  this.iterations = config.iterations || DEFAULT_ITERATIONS;
  this.keyLength = config.keyLength || DEFAULT_KEY_LENGTH;
  this.digest = config.digest || DEFAULT_DIGEST;
  this.defaultSecret = config.secret;
  this.context = config.context || {};
}

Krypt.prototype.DEFAULT_DIGEST = DEFAULT_DIGEST;
Krypt.prototype.DEFAULT_DEPRECATED_DIGEST = DEFAULT_DEPRECATED_DIGEST;

Krypt.prototype.setSecret = function setSecret(secret) {
  this.defaultSecret = secret;
};


Krypt.prototype.setIterations = function setIterations(iterations) {
  this.iterations = iterations;
};


Krypt.prototype.setKeyLength = function setKeyLength(keyLength) {
  this.keyLength = keyLength;
};


Krypt.prototype.setContext = function setContext(context) {
  this.context = context;
};

Krypt.prototype.setDigest = function setDigest(digest) {
  this.digest = digest;
};


Krypt.prototype.encrypt = function encrypt(input, secret, cb) {

  // Support both async and sync encryption.
  // This has implications primarily in the key stretching as it's very expensive.
  var async = (cb && typeof cb === 'function') || false;

  var self = this;

  if (!input) {
    var inputError = new Error('You must provide a value to encrypt');
    if (async) {
      return cb(inputError);
    } else {
      throw inputError;
    }
  }

  secret = secret || this.defaultSecret;
  if (!secret) {
    var secretError = new Error('A \'secret\' is required to encrypt');
    if (async) {
      return cb(secretError);
    } else {
      throw secretError;
    }
  }

  // Legacy check to deal with old versions that recorded key length in Bytes
  if (this.keyLength === 32) {
    this.keyLength = this.keyLength * 8;
  }

  var salt = crypto.randomBytes(this.keyLength / 8),
      iv = crypto.randomBytes(16);

  function encryptUsingKey(key) {
    var cipher = crypto.createCipheriv(CIPHER, key, iv);

    var encryptedValue = cipher.update(input, 'utf8', 'base64');
    encryptedValue += cipher.final('base64');

    var result = {
      cipher: CIPHER,
      keyDerivation: KEY_DERIVATION,
      keyLength: self.keyLength,
      iterations: self.iterations,
      iv: iv.toString('base64'),
      salt: salt.toString('base64'),
      digest: self.digest,
      value: encryptedValue
    };

    _.defaults(result, self.context);

    return result;
  }

  if (async) {
    crypto.pbkdf2(secret, salt, this.iterations, this.keyLength / 8, this.digest, function (err, key) {
      if (err) {
        return cb(err);
      }

      try {
        var result = encryptUsingKey(key);
        cb(null, result);
      } catch (err) {
        cb(err);
      }
    });
  } else {
    try {
      var key = crypto.pbkdf2Sync(secret, salt, this.iterations, this.keyLength / 8, this.digest);
      return encryptUsingKey(key);
    } catch (err) {
      throw new Error('Unable to encrypt value due to: ' + err);
    }
  }
};


Krypt.prototype.decrypt = function decrypt(input, secret, cb) {

  // Support both async and sync encryption.
  // This has implications primarily in the key stretching as it's very expensive.
  var async = (cb && typeof cb === 'function') || false;

  // Ensure we have something to decrypt
  if (!input) {
    var inputError = new Error('You must provide a value to decrypt');
    if (async) {
      return cb(inputError);
    } else {
      throw inputError;
    }
  }

  // Ensure we have the secret used to encrypt this value
  secret = secret || this.defaultSecret;
  if (!secret) {
    var secretError = new Error('A \'secret\' is required to decrypt');
    if (async) {
      return cb(secretError);
    } else {
      throw secretError;
    }
  }

  // If we get a string as input, turn it into an object
  if (typeof input !== 'object') {
    try {
      input = JSON.parse(input);
    } catch (err) {
      var parseError = new Error('Unable to parse string input as JSON');
      if (async) {
        return cb(parseError);
      } else {
        throw parseError;
      }
    }
  }

  // Ensure our input is a valid object with 'iv', 'salt', and 'value'
  if (!input.iv || !input.salt || !input.value) {
    var validationError = new Error('Input must be a valid object with \'iv\', \'salt\', and \'value\' properties');
    if (async) {
      return cb(validationError);
    } else {
      throw validationError;
    }
  }

  var salt = new Buffer(input.salt, 'base64'),
      iv = new Buffer(input.iv, 'base64'),
      keyLength = input.keyLength || this.keyLength,
      iterations = input.iterations || this.iterations,
      digest = input.digest || DEFAULT_DEPRECATED_DIGEST;

  // Legacy check to deal with old versions that recorded key length in Bytes
  if (keyLength === 32) {
    keyLength = keyLength * 8;
  }

  function decryptUsingKey(key) {
    var decipher = crypto.createDecipheriv(CIPHER, key, iv);

    var decryptedValue = decipher.update(input.value, 'base64', 'utf8');
    decryptedValue += decipher.final('utf8');

    return decryptedValue;
  }

  if (async) {
    crypto.pbkdf2(secret, salt, iterations, keyLength / 8, digest, function (err, key) {
      if (err) {
        return cb(err);
      }

      try {
        var result = decryptUsingKey(key);
        cb(null, result);
      } catch (err) {
        cb(err);
      }
    });
  } else {
    try {
      var key = crypto.pbkdf2Sync(secret, salt, iterations, keyLength / 8, digest);
      return decryptUsingKey(key);
    } catch (err) {
      throw new Error('Unable to decrypt value due to: ' + err);
    }
  }
};


Krypt.prototype.encryptAsync = function encryptAsync(input, secret, cb) {
  this.encrypt(input, secret, cb);
};


Krypt.prototype.decryptAsync = function decryptAsync(input, secret, cb) {
  this.decrypt(input, secret, cb);
};


Krypt.prototype.encryptSync = function encryptSync(input, secret) {
  return this.encrypt(input, secret);
};


Krypt.prototype.decryptSync = function decryptSync(input, secret) {
  return this.decrypt(input, secret);
};
